### DISCUSSION TITLE
Code for analog fuel gauge voltage measurement

### DISCUSSION THREAD

**Discussion by ESchlotfeldt** [hatlabs/discussions#124] [Q&A]
I want to configure the HALMET sample software to be able to read the voltage off the tank sender in parallel to the existing analog Fuel gauge. Seems to me that the sample code for the tank level calc (in halmet_analog.cpp) is set up to measure the resistance of the sender only. Has anyone got some sample code for the Halmet to do a voltage measurement, that I can examine and adapt in order to achieve that?

**Comment by mairas** [comment]
Hey, sorry about the delay.

This was a clear omission in the example code, so I added a reading a voltage input level there. See the `dev-3` branch `main.cpp` from line 234 onwards.

Furthermore, I noticed the analog input scale factors were off. Fixed them.

Finally, I also noted that the analog resistance and/or tank level Signal K paths were broken in the example repo, so they're fixed as well.

**Reply by ESchlotfeldt** [reply]
Thank you - I had modified your resistance sender code to give me a voltage reading (I found kAnalogInputScale =10 by trial and error):

auto sender_voltage =
      new RepeatSensor<float>(ads_read_delay, [ads1115, channel]() {
        int16_t adc_output = ads1115->readADC_SingleEnded(channel);
        float adc_output_volts = ads1115->computeVolts(adc_output);
      return kAnalogInputScale * adc_output_volts;  
      });

This seems to work well, but I see that you seem to have tackled the issue quite differently - is there a problem with doing it my way?

**Reply by mairas** [reply]
Your approach is perfectly fine. I created the dedicated class to allow UI configuration of the calibration value.

**Reply by ESchlotfeldt** [reply]
Ah OK, thanks!

**Reply by ESchlotfeldt** [reply]
I've managed to get all the measurements working on the Halmet - great piece of hardware!

I need some advice however on the following 2 issues:

1. I'm stuck on what should be a really simple piece of coding - working out "fuel consumption" and "engine hours remaining", based on "RPM" and "fuel tank level". I've previously had this working on my esp32, where I had all the code in main.cpp as per the standard sample code, but because I'm not so clued up on OOP I am struggling to implement it here trying to pass the fuel tank level output from digital.cpp to a new subroutine called endurance.cpp via main.cpp. I'd appreciate suggestions on how to do that. 

2. I then want to calculate "range remaining" from "engine hours remaining" and "SOG". Can I do this in the HALMET code? If so how do I get access to SOG to perform this calculation? If not, how could I do this in SignalK?

Appreciate any help and guidance!

**Reply by mairas** [reply]
At the moment, you need to receive the multiple inputs into a variable and then have another method that takes the variable values, computes the resulting value and emits this. The emitting thing can be e.g. an onRepeat reaction that checks the runs every X milliseconds. Have a look at the n2k_senders.h class implementations: they have LambdaConsumers that take values and then update member variables based on that. Similarly, they have onRepeat reactions that construct NMEA 2000 sentences and transmit them at regular intervals.

This is a bit low level and repetitive, though, and SensESP is lacking good transforms for combining data. I'll have a look.

**Reply by ESchlotfeldt** [reply]
Thanks for the prompt response. Had a look - seems a bit tricky for my level of expertise, maybe I'll wait for the new transforms you mention. In the interim I think I've found a way to calculate this in signal K through the "signalk-derived-data" plugin, which seems to allow users the opportunity to add custom calculations to the plugin.

**Reply by mairas** [reply]
Here's a PR that includes transforms that should help you: https://github.com/SignalK/SensESP/pull/703

The `Join` and `Zip` transforms take multiple values and combine them into a single tuple that can be processed together. I wrote an example that demonstrates both concepts. With either of them, you can combine "RPM" and "tank_level" and process them together in a single LambdaTransform.

I'll let the PR sit for a few days in case someone would be interested in reviewing it. You can already run the code by specifying the `new_transforms` branch in your `platformio.ini` `lib_deps`.
